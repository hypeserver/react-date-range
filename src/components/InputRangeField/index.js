import PropTypes from 'prop-types';
import React, { Component } from 'react';

const MIN = 0;
const MAX = 99999;

class InputRangeField extends Component {
  constructor(props, context) {
    super(props, context);
  }

  shouldComponentUpdate(nextProps) {
    const { value, label, placeholder } = this.props;

    return value !== nextProps.value || label !== nextProps.label || placeholder !== nextProps.placeholder;
  }

  onChange = e => {
    const { onChange } = this.props;

    let value = parseInt(e.target.value, 10);
    value = isNaN(value) ? 0 : Math.max(Math.min(MAX, value), MIN);

    onChange(value);
  };

  render() {
    const { label, placeholder, value, styles, onBlur, onFocus } = this.props;

    return (
      <div className={styles.inputRange}>
        <input
          className={styles.inputRangeInput}
          placeholder={placeholder}
          value={value}
          min={MIN}
          max={MAX}
          onChange={this.onChange}
          onFocus={onFocus}
          onBlur={onBlur}
        />
        <span className={styles.inputRangeLabel}>{label}</span>
      </div>
    );
  }
}

InputRangeField.propTypes = {
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  label: PropTypes.oneOfType([PropTypes.element, PropTypes.node]).isRequired,
  placeholder: PropTypes.string,
  styles: PropTypes.shape({
    inputRange: PropTypes.string,
    inputRangeInput: PropTypes.string,
    inputRangeLabel: PropTypes.string,
  }).isRequired,
  onBlur: PropTypes.func.isRequired,
  onFocus: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
};

InputRangeField.defaultProps = {
  value: '',
  placeholder: '-',
};

export default InputRangeField;
