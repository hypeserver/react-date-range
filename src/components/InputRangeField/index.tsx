import React, { ChangeEvent, Component, FocusEventHandler } from 'react';

const MIN = 0;
const MAX = 99999;

type Styles = {
  inputRange: string;
  inputRangeInput: string;
  inputRangeLabel: string;
}

type ComponentProps = {
  value: string | number;
  label: string | JSX.Element,
  placeholder: string,
  styles: Styles;
  onBlur: FocusEventHandler<HTMLInputElement>;
  onFocus: FocusEventHandler<HTMLInputElement>;
  onChange: (n: number) => void;
};


class InputRangeField extends Component<ComponentProps> {
  public static defaultProps = {
    value: '',
    placeholder: '-',
  };

  constructor(props: ComponentProps) {
    super(props);
  }

  shouldComponentUpdate(nextProps: ComponentProps) {
    const { value, label, placeholder } = this.props;

    return (
      value !== nextProps.value ||
      label !== nextProps.label ||
      placeholder !== nextProps.placeholder
    );
  }

  onChange = (e: ChangeEvent<HTMLInputElement>) => {
    const value = parseInt(e.target.value, 10);
    const changeArg = isNaN(value) ? 0 : Math.max(Math.min(MAX, value), MIN);
    this.props.onChange(changeArg);
  };

  render() {
    const { label, placeholder, value, styles, onBlur, onFocus } = this.props;

    return (
      <div className={styles.inputRange}>
        <input
          className={styles.inputRangeInput}
          placeholder={placeholder}
          value={value}
          min={MIN}
          max={MAX}
          onChange={this.onChange}
          onFocus={onFocus}
          onBlur={onBlur}
        />
        <span className={styles.inputRangeLabel}>{label}</span>
      </div>
    );
  }
}


export default InputRangeField;
