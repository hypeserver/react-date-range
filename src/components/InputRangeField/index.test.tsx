import React from 'react';
import { mount } from 'enzyme';

import InputRangeField from '.';

const styles = {
  inputRange: 'range',
  inputRangeInput: 'input',
  inputRangeLabel: 'label',
};
const toChangeEvent = (value: string | number) => ({ target: { value } });

describe('InputRangeField tests', () => {
  test('Should parse input value to number', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <InputRangeField
        label="Input label"
        styles={styles}
        onChange={onChange}
        onFocus={jest.fn()}
        onBlur={jest.fn()}
      />
    );

    wrapper.find('input').simulate('change', toChangeEvent('3'));
    expect(onChange).lastCalledWith(3);
    wrapper.find('input').simulate('change', toChangeEvent(12));
    expect(onChange).lastCalledWith(12);
    wrapper.find('input').simulate('change', toChangeEvent(''));
    expect(onChange).lastCalledWith(0);
    wrapper.find('input').simulate('change', toChangeEvent('invalid number'));
    expect(onChange).lastCalledWith(0);
    wrapper.find('input').simulate('change', toChangeEvent(-12));
    expect(onChange).lastCalledWith(0);
    wrapper.find('input').simulate('change', toChangeEvent(99999999));
    expect(onChange).lastCalledWith(99999);

    expect(onChange).toHaveBeenCalledTimes(6);
    expect(wrapper).toMatchSnapshot();
  });

  test('Should rerender when props change', () => {
    const wrapper = mount(
      <InputRangeField
        value={12}
        placeholder="Placeholder"
        label="Input label"
        styles={styles}
        onChange={jest.fn()}
        onFocus={jest.fn()}
        onBlur={jest.fn()}
      />
    );

    expect(wrapper.find(`.${styles.inputRangeInput}`).prop('value')).toEqual(12);
    expect(wrapper.find(`.${styles.inputRangeInput}`).prop('placeholder')).toEqual('Placeholder');
    expect(wrapper.find(`.${styles.inputRangeLabel}`).text()).toEqual('Input label');

    wrapper.setProps({ value: '32' });
    expect(wrapper.find(`.${styles.inputRangeInput}`).prop('value')).toEqual('32');
    expect(wrapper.find(`.${styles.inputRangeInput}`).prop('placeholder')).toEqual('Placeholder');
    expect(wrapper.find(`.${styles.inputRangeLabel}`).text()).toEqual('Input label');

    wrapper.setProps({ placeholder: '-' });
    expect(wrapper.find(`.${styles.inputRangeInput}`).prop('value')).toEqual('32');
    expect(wrapper.find(`.${styles.inputRangeInput}`).prop('placeholder')).toEqual('-');
    expect(wrapper.find(`.${styles.inputRangeLabel}`).text()).toEqual('Input label');

    wrapper.setProps({ label: 'Label' });
    expect(wrapper.find(`.${styles.inputRangeInput}`).prop('value')).toEqual('32');
    expect(wrapper.find(`.${styles.inputRangeInput}`).prop('placeholder')).toEqual('-');
    expect(wrapper.find(`.${styles.inputRangeLabel}`).text()).toEqual('Label');
  });

  test('Should render the label as a Component', () => {
    const Label = () => <span className="input-range-field-label">Input label</span>;
    const wrapper = mount(
      <InputRangeField
        value={12}
        placeholder="Placeholder"
        label={<Label />}
        styles={styles}
        onChange={jest.fn()}
        onFocus={jest.fn()}
        onBlur={jest.fn()}
      />
    );

    expect(wrapper.find(`.${styles.inputRangeInput}`).prop('value')).toEqual(12);
    expect(wrapper.find(`.${styles.inputRangeInput}`).prop('placeholder')).toEqual('Placeholder');
    expect(wrapper.find(`.${styles.inputRangeLabel}`).text()).toEqual('Input label');
    expect(wrapper.find(`.${styles.inputRangeLabel}`).text()).toEqual('Input label');
    expect(wrapper.find(`.input-range-field-label`).text()).toEqual('Input label');

    wrapper.setProps({ value: '32' });
    expect(wrapper.find(`.${styles.inputRangeInput}`).prop('value')).toEqual('32');
    expect(wrapper.find(`.${styles.inputRangeInput}`).prop('placeholder')).toEqual('Placeholder');
    expect(wrapper.find(`.${styles.inputRangeLabel}`).text()).toEqual('Input label');

    wrapper.setProps({ placeholder: '-' });
    expect(wrapper.find(`.${styles.inputRangeInput}`).prop('value')).toEqual('32');
    expect(wrapper.find(`.${styles.inputRangeInput}`).prop('placeholder')).toEqual('-');
    expect(wrapper.find(`.${styles.inputRangeLabel}`).text()).toEqual('Input label');

    wrapper.setProps({ label: 'Label' });
    expect(wrapper.find(`.${styles.inputRangeInput}`).prop('value')).toEqual('32');
    expect(wrapper.find(`.${styles.inputRangeInput}`).prop('placeholder')).toEqual('-');
    expect(wrapper.find(`.${styles.inputRangeLabel}`).text()).toEqual('Label');
  });
});
