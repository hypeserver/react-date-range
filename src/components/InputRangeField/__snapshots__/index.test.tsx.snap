// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputRangeField tests Should parse input value to number 1`] = `
<InputRangeField
  label="Input label"
  onBlur={[MockFunction]}
  onChange={
    [MockFunction] {
      "calls": Array [
        Array [
          3,
        ],
        Array [
          12,
        ],
        Array [
          0,
        ],
        Array [
          0,
        ],
        Array [
          0,
        ],
        Array [
          99999,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  onFocus={[MockFunction]}
  placeholder="-"
  styles={
    Object {
      "inputRange": "range",
      "inputRangeInput": "input",
      "inputRangeLabel": "label",
    }
  }
  value=""
>
  <div
    className="range"
  >
    <input
      className="input"
      max={99999}
      min={0}
      onBlur={[MockFunction]}
      onChange={[Function]}
      onFocus={[MockFunction]}
      placeholder="-"
      value=""
    />
    <span
      className="label"
    >
      Input label
    </span>
  </div>
</InputRangeField>
`;
